## 1. Hava durumu modeli

In [1]:
import numpy as np
import pandas as pd
import seaborn as sns
import statsmodels.api as sm
from sklearn import linear_model
import matplotlib.pyplot as plt
import seaborn as sns
from sklearn import preprocessing
from scipy.stats.mstats import winsorize
import warnings

In [2]:
df = pd.read_csv('weather.csv')
df.head()

Formatted Date        Summary Precip Type  Temperature (C)  \
0  2006-04-01 00:00:00.000 +0200  Partly Cloudy        rain         9.472222   
1  2006-04-01 01:00:00.000 +0200  Partly Cloudy        rain         9.355556   
2  2006-04-01 02:00:00.000 +0200  Mostly Cloudy        rain         9.377778   
3  2006-04-01 03:00:00.000 +0200  Partly Cloudy        rain         8.288889   
4  2006-04-01 04:00:00.000 +0200  Mostly Cloudy        rain         8.755556   

   Apparent Temperature (C)  Humidity  Wind Speed (km/h)  \
0                  7.388889      0.89            14.1197   
1                  7.227778      0.86            14.2646   
2                  9.377778      0.89             3.9284   
3                  5.944444      0.83            14.1036   
4                  6.977778      0.83            11.0446   

   Wind Bearing (degrees)  Visibility (km)  Loud Cover  Pressure (millibars)  \
0                   251.0          15.8263         0.0               1015.13   
1                   259.0          15.8263         0.0               1015.63   
2                   204.0          14.9569         0.0               1015.94   
3                   269.0          15.8263         0.0               1016.41   
4                   259.0          15.8263         0.0               1016.51   

                       Daily Summary  
0  Partly cloudy throughout the day.  
1  Partly cloudy throughout the day.  
2  Partly cloudy throughout the day.  
3  Partly cloudy throughout the day.  
4  Partly cloudy throughout the day.

In [3]:
Y = df['Temperature (C)'] - df['Apparent Temperature (C)']
X = df[['Humidity','Wind Speed (km/h)']]
X = sm.add_constant(X)

results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.288
Model:                            OLS   Adj. R-squared:                  0.288
Method:                 Least Squares   F-statistic:                 1.949e+04
Date:                Tue, 21 May 2019   Prob (F-statistic):               0.00
Time:                        18:18:42   Log-Likelihood:            -1.7046e+05
No. Observations:               96453   AIC:                         3.409e+05
Df Residuals:                   96450   BIC:                         3.409e+05
Df Model:                           2                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -2.4381      0.021   -115.948      0.000      -2.479      -2.397
Humidity              3.0292      0.024    126.479      0.000       2.982       3.076
Wind Speed (km/h)     0.1193      0.001    176.164      0.000       0.118       0.121
==============================================================================
Omnibus:                     3935.747   Durbin-Watson:                   0.264
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4613.311
Skew:                           0.478   Prob(JB):                         0.00
Kurtosis:                       3.484   Cond. No.                         88.1
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R-kare ve ayarlanmış R-kare değerleri birbirine eşit ve 0.288'dir.R-kare değerinin overfitting riski taşımadan 1' yakın olmasını isteriz
bu sebepten 0.288 değeri bizim için düşük bir sonuçtur

### Yeni modelimiz

In [4]:
df["humidity_windspeed"]=df["Humidity"]*df['Wind Speed (km/h)']
X = df[["Humidity","Wind Speed (km/h)","humidity_windspeed"]]

X = sm.add_constant(X)

results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.341
Model:                            OLS   Adj. R-squared:                  0.341
Method:                 Least Squares   F-statistic:                 1.666e+04
Date:                Tue, 21 May 2019   Prob (F-statistic):               0.00
Time:                        18:18:42   Log-Likelihood:            -1.6669e+05
No. Observations:               96453   AIC:                         3.334e+05
Df Residuals:                   96449   BIC:                         3.334e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
======================================================================================
                         coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------------
const                 -0.0839      0.033     -2.511      0.012      -0.149      -0.018
Humidity              -0.1775      0.043     -4.133      0.000      -0.262      -0.093
Wind Speed (km/h)     -0.0905      0.002    -36.797      0.000      -0.095      -0.086
humidity_windspeed     0.2971      0.003     88.470      0.000       0.291       0.304
==============================================================================
Omnibus:                     4849.937   Durbin-Watson:                   0.262
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             9295.404
Skew:                           0.378   Prob(JB):                         0.00
Kurtosis:                       4.320   Cond. No.                         193.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R-kare ve ayarlanmış R-kare sonuçlarımız yine aynı ve 0.341 olarak değişmekte olup bir artış söz konusudur.
2. modelimiz ilk modele göre daha iyi bir konumdadır.

#### Görünürlük eklenmiş model

In [5]:
X = df[['Humidity','Wind Speed (km/h)', 'Visibility (km)']]

X = sm.add_constant(X)

results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                      y   R-squared:                       0.304
Model:                            OLS   Adj. R-squared:                  0.303
Method:                 Least Squares   F-statistic:                 1.401e+04
Date:                Tue, 21 May 2019   Prob (F-statistic):               0.00
Time:                        18:18:44   Log-Likelihood:            -1.6938e+05
No. Observations:               96453   AIC:                         3.388e+05
Df Residuals:                   96449   BIC:                         3.388e+05
Df Model:                           3                                         
Covariance Type:            nonrobust                                         
=====================================================================================
                        coef    std err          t      P>|t|      [0.025      0.975]
-------------------------------------------------------------------------------------
const                -1.5756      0.028    -56.605      0.000      -1.630      -1.521
Humidity              2.6066      0.025    102.784      0.000       2.557       2.656
Wind Speed (km/h)     0.1199      0.001    179.014      0.000       0.119       0.121
Visibility (km)      -0.0540      0.001    -46.614      0.000      -0.056      -0.052
==============================================================================
Omnibus:                     3833.895   Durbin-Watson:                   0.279
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             4584.022
Skew:                           0.459   Prob(JB):                         0.00
Kurtosis:                       3.545   Cond. No.                         131.
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
"""

R-kare değerimiz 0.304 ayarlanmış R-kare değeri 0.303'tür.R-kare değeri model 3'te model 2'ye göre 0.037'lik bir düşüşe sahiptir.
Yani R-kare açısında model 2 daha iyi gözükmektedir.

AIC ve BIC değerleri açısından;
1. model için AIC :3.409e+05 BIC:3.409e+05
2. model için AIC :3.334e+05 BIC:3.334e+05
3. model için AIC :3.388e+05 BIC:3.388e+05
AIC ve BIC değeri en düşük olan 2. model en iyi modeldir.Sonrasında 3.model ve son sırada 1. model gelmektedir.

## 2. Ev fiyatları modeli

In [14]:
df3 = pd.read_csv('train.csv')
df3.head()

Id  MSSubClass MSZoning  LotFrontage  LotArea Street Alley LotShape  \
0   1          60       RL         65.0     8450   Pave   NaN      Reg   
1   2          20       RL         80.0     9600   Pave   NaN      Reg   
2   3          60       RL         68.0    11250   Pave   NaN      IR1   
3   4          70       RL         60.0     9550   Pave   NaN      IR1   
4   5          60       RL         84.0    14260   Pave   NaN      IR1   

  LandContour Utilities    ...     PoolArea PoolQC Fence MiscFeature MiscVal  \
0         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
1         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
2         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
3         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   
4         Lvl    AllPub    ...            0    NaN   NaN         NaN       0   

  MoSold YrSold  SaleType  SaleCondition  SalePrice  
0      2   2008        WD         Normal     208500  
1      5   2007        WD         Normal     181500  
2      9   2008        WD         Normal     223500  
3      2   2006        WD        Abnorml     140000  
4     12   2008        WD         Normal     250000  

[5 rows x 81 columns]

In [15]:
df3 = df3.drop(['Alley', 'PoolQC', 'Fence','MiscFeature','LotFrontage', 'FireplaceQu' ], axis = 1)
df3 = df3.dropna()

In [16]:
Y = df3["SalePrice"]
X = df3[["MSSubClass","LotArea","MoSold","PoolArea"]]
X = sm.add_constant(X)

results = sm.OLS(Y,X).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.074
Model:                            OLS   Adj. R-squared:                  0.071
Method:                 Least Squares   F-statistic:                     26.73
Date:                Tue, 21 May 2019   Prob (F-statistic):           2.34e-21
Time:                        18:19:30   Log-Likelihood:                -16934.
No. Observations:                1338   AIC:                         3.388e+04
Df Residuals:                    1333   BIC:                         3.390e+04
Df Model:                           4                                         
Covariance Type:            nonrobust                                         
==============================================================================
                 coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------
const       1.636e+05   6571.224     24.892      0.000    1.51e+05    1.76e+05
MSSubClass   -90.1196     50.889     -1.771      0.077    -189.951       9.712
LotArea        1.8524      0.204      9.095      0.000       1.453       2.252
MoSold      1262.5576    770.975      1.638      0.102    -249.899    2775.014
PoolArea     141.4281     49.750      2.843      0.005      43.831     239.025
==============================================================================
Omnibus:                      535.818   Durbin-Watson:                   2.041
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             2863.259
Skew:                           1.800   Prob(JB):                         0.00
Kurtosis:                       9.197   Cond. No.                     4.72e+04
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 4.72e+04. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

In [17]:
df3.corr()

Id  MSSubClass   LotArea  OverallQual  OverallCond  \
Id             1.000000    0.016725 -0.036083    -0.044455     0.011850   
MSSubClass     0.016725    1.000000 -0.134158     0.051690    -0.098153   
LotArea       -0.036083   -0.134158  1.000000     0.090451     0.001415   
OverallQual   -0.044455    0.051690  0.090451     1.000000    -0.163080   
OverallCond    0.011850   -0.098153  0.001415    -0.163080     1.000000   
YearBuilt     -0.023595    0.056456 -0.005022     0.570668    -0.417330   
YearRemodAdd  -0.028434    0.038513  0.001681     0.549460     0.035076   
MasVnrArea    -0.051038    0.031333  0.096727     0.401175    -0.141080   
BsmtFinSF1    -0.004919   -0.069053  0.210148     0.197441    -0.060068   
BsmtFinSF2    -0.004908   -0.079329  0.110493    -0.086014     0.035698   
BsmtUnfSF     -0.010196   -0.141025 -0.014589     0.292940    -0.161038   
TotalBsmtSF   -0.018637   -0.263613  0.267188     0.505897    -0.228020   
1stFlrSF       0.009750   -0.258787  0.288502     0.481978    -0.144066   
2ndFlrSF      -0.002866    0.306709  0.047773     0.281145     0.011685   
LowQualFinSF  -0.047958    0.019990  0.009729    -0.016704     0.041865   
GrLivArea      0.001067    0.068782  0.255445     0.594470    -0.093843   
BsmtFullBath   0.010210   -0.011317  0.160687     0.083254    -0.070089   
BsmtHalfBath  -0.028333   -0.000151  0.046420    -0.064952     0.119671   
FullBath      -0.000334    0.117720  0.115727     0.580425    -0.211882   
HalfBath      -0.013197    0.197520  0.006052     0.249283    -0.073568   
BedroomAbvGr   0.041290   -0.065699  0.120369     0.081994     0.023180   
KitchenAbvGr   0.013834    0.239177 -0.012611    -0.135084    -0.070910   
TotRmsAbvGrd   0.022852    0.014514  0.185654     0.442941    -0.063883   
Fireplaces    -0.017839   -0.024752  0.263061     0.377359    -0.031910   
GarageYrBlt   -0.003114    0.082040 -0.025700     0.545835    -0.336541   
GarageCars     0.009797    0.006872  0.138228     0.591878    -0.249832   
GarageArea     0.013658   -0.070278  0.168630     0.530092    -0.204905   
WoodDeckSF    -0.032801   -0.010175  0.167822     0.218570    -0.012947   
OpenPorchSF   -0.005332   -0.000359  0.084942     0.307615    -0.060429   
EnclosedPorch  0.005106   -0.022598 -0.017968    -0.116931     0.071473   
3SsnPorch     -0.043766   -0.041391  0.019326     0.026470     0.023789   
ScreenPorch    0.000904   -0.022748  0.038489     0.045257     0.053759   
PoolArea       0.059455    0.010199  0.077050     0.064570    -0.003544   
MiscVal       -0.002206   -0.021512  0.040824    -0.029351     0.077314   
MoSold         0.020939   -0.016283  0.000118     0.062641    -0.011772   
YrSold         0.002241   -0.021591 -0.012066    -0.013949     0.056340   
SalePrice     -0.030473   -0.079599  0.254757     0.783546    -0.108627   

               YearBuilt  YearRemodAdd  MasVnrArea  BsmtFinSF1  BsmtFinSF2  \
Id             -0.023595     -0.028434   -0.051038   -0.004919   -0.004908   
MSSubClass      0.056456      0.038513    0.031333   -0.069053   -0.079329   
LotArea        -0.005022      0.001681    0.096727    0.210148    0.110493   
OverallQual     0.570668      0.549460    0.401175    0.197441   -0.086014   
OverallCond    -0.417330      0.035076   -0.141080   -0.060068    0.035698   
YearBuilt       1.000000      0.617156    0.299410    0.224264   -0.070403   
YearRemodAdd    0.617156      1.000000    0.168311    0.099681   -0.085268   
MasVnrArea      0.299410      0.168311    1.000000    0.253015   -0.083303   
BsmtFinSF1      0.224264      0.099681    0.253015    1.000000   -0.064855   
BsmtFinSF2     -0.070403     -0.085268   -0.083303   -0.064855    1.000000   
BsmtUnfSF       0.147762      0.171881    0.104662   -0.552809   -0.233449   
TotalBsmtSF     0.385146      0.264281    0.365629    0.505000    0.083184   
1stFlrSF        0.263585      0.241889    0.341001    0.450366    0.094065   
2ndFlrSF        0.006989      0.120545    0.169730   -0.151876   -0.109254   


In [18]:
X = X.drop(["MoSold","MSSubClass"], axis = 1)
X.head()

const  LotArea  PoolArea
0    1.0     8450         0
1    1.0     9600         0
2    1.0    11250         0
3    1.0     9550         0
4    1.0    14260         0

In [24]:
X3 = df3[["LotArea", "PoolArea", "OverallQual","YearRemodAdd","YearBuilt"]]

X3= sm.add_constant(X3)
results = sm.OLS(Y,X3).fit()
results.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:              SalePrice   R-squared:                       0.660
Model:                            OLS   Adj. R-squared:                  0.659
Method:                 Least Squares   F-statistic:                     517.8
Date:                Tue, 21 May 2019   Prob (F-statistic):          3.61e-309
Time:                        18:22:27   Log-Likelihood:                -16263.
No. Observations:                1338   AIC:                         3.254e+04
Df Residuals:                    1332   BIC:                         3.257e+04
Df Model:                           5                                         
Covariance Type:            nonrobust                                         
================================================================================
                   coef    std err          t      P>|t|      [0.025      0.975]
--------------------------------------------------------------------------------
const        -1.097e+06   1.49e+05     -7.343      0.000   -1.39e+06   -8.04e+05
LotArea          1.4575      0.123     11.844      0.000       1.216       1.699
PoolArea        61.9274     30.202      2.050      0.041       2.678     121.177
OverallQual   4.049e+04   1228.492     32.959      0.000    3.81e+04    4.29e+04
YearRemodAdd   346.2957     82.923      4.176      0.000     183.623     508.969
YearBuilt      166.6718     57.976      2.875      0.004      52.938     280.405
==============================================================================
Omnibus:                      559.375   Durbin-Watson:                   1.964
Prob(Omnibus):                  0.000   Jarque-Bera (JB):             8081.063
Skew:                           1.542   Prob(JB):                         0.00
Kurtosis:                      14.638   Cond. No.                     1.78e+06
==============================================================================

Warnings:
[1] Standard Errors assume that the covariance matrix of the errors is correctly specified.
[2] The condition number is large, 1.78e+06. This might indicate that there are
strong multicollinearity or other numerical problems.
"""

F-testi:?
    
R-kare :İlk model için 0.074 olan R-kare değeri 0.660 olmaktadır.Bunun sebebi 2. modelde hedef değişkenin yüksek korelasyon ilişkisi
içinde olduğu değerler ile çalışmamızdır.Bu yüzden 2.model daha iyidir.

Ayarlanmış R-kare:Ayarlanmış R-kare değerimiz 0.071'den 0.659'a yükselmiştir.Yine aynı sebepten ötürü 2. model daha iyi gözükmektedir.

AIC ve BIC:3.388e+04 ve 3.254e+04 sırayla AIC değerlerimizdir.AIC değerinin düşüklüğü istenen bir şeydir.Bu da model 2'de daha verimli 
bir hale dönüşmüştür.
    